subsys/bluetooth/host/gatt.c

```diff
@@ -3606,37 +3606,46 @@ static bool check_subscribe_security_level(struct bt_conn *conn,
 	return true;
 }
 
-void bt_gatt_notification(struct bt_conn *conn, uint16_t handle,
-			  const void *data, uint16_t length)
+static void call_notify_cb_and_maybe_unsubscribe(struct bt_conn *conn, struct gatt_sub *sub,
+						 uint16_t handle, const void *data, uint16_t length)
 {
 	struct bt_gatt_subscribe_params *params, *tmp;
-	struct gatt_sub *sub;
-
-	LOG_DBG("handle 0x%04x length %u", handle, length);
-
-	sub = gatt_sub_find(conn);
-	if (!sub) {
-		return;
-	}
+	int err;
 
 	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&sub->list, params, tmp, node) {
 		if (handle != params->value_handle) {
 			continue;
 		}
 
 		if (check_subscribe_security_level(conn, params)) {
-			if (params->notify(conn, params, data, length) ==
-			    BT_GATT_ITER_STOP) {
-				bt_gatt_unsubscribe(conn, params);
+			if (params->notify(conn, params, data, length) == BT_GATT_ITER_STOP) {
+				err = bt_gatt_unsubscribe(conn, params);
```

Comment:
Not related to this PR, but I have to say that this looks a bit awkward as an API, since it seems like it'd be trivial (and more clear & explicit) for the callback itself to do this `unsubscribe()` call rather than implying it through the return value.

Reply:
I agree. I spelled it out in the function name to surface that awkwardness :)

Reply:
I also agree - Maybe something we can avoid if we get a new GATT API for EATT